src/client/discard_task.cpp

```diff
@@ -0,0 +1,148 @@
+                        "filename = "
+                     << fileInfo->fullPathName << ", offset = " << offset
+                     << ", taskid = " << timerId_;
+        metric_->totalCanceled << 1;
+        fileSegment->ReleaseLock();
+        taskManager_->OnTaskFinish(timerId_);
+        return;
+    }
+
+    LIBCURVE_ERROR errCode = mdsClient_->DeAllocateSegment(fileInfo, offset);
+    if (errCode == LIBCURVE_ERROR::OK) {
+        metric_->totalSuccess << 1;
+        fileSegment->ClearBitmap();
+        metaCache_->CleanChunksInSegment(segmentIndex_);
+        LOG(INFO) << "DiscardTask success, filename = "
+                  << fileInfo->fullPathName << ", offset = " << offset
+                  << ", taskid = " << timerId_;
+    } else if (errCode == LIBCURVE_ERROR::UNDER_SNAPSHOT) {
+        metric_->totalError << 1;
+        LOG(WARNING) << "DiscardTask failed, " << fileInfo->fullPathName
+                     << " has snapshot, offset = " << offset
+                     << ", taskid = " << timerId_;
+    } else {
+        metric_->totalError << 1;
+        LOG(ERROR) << "DiscardTask failed, mds return error = " << errCode
+                   << ", filename = " << fileInfo->fullPathName
+                   << ", offset = " << offset << ", taskid = " << timerId_;
+    }
+
+    fileSegment->ReleaseLock();
+    taskManager_->OnTaskFinish(timerId_);
+}
+
+DiscardTaskManager::DiscardTaskManager(DiscardMetric* metric)
+    : mtx_(), cond_(), unfinishedTasks_(), metric_(metric) {}
+
+static void RunDiscardTask(void* arg) {
+    DiscardTask* task = static_cast<DiscardTask*>(arg);
+    task->Run();
+}
+
+void DiscardTaskManager::OnTaskFinish(bthread_timer_t timerId) {
+    std::lock_guard<bthread::Mutex> lk(mtx_);
+    unfinishedTasks_.erase(timerId);
+    cond_.notify_one();
+}
+
+bool DiscardTaskManager::ScheduleTask(SegmentIndex segmentIndex,
+                                      MetaCache* metaCache,
+                                      MDSClient* mdsclient, timespec abstime) {
+    bthread_timer_t timerId;
+    std::unique_ptr<DiscardTask> task(
+        new DiscardTask(this, segmentIndex, metaCache, mdsclient, metric_));
+
+    int ret = bthread_timer_add(&timerId, abstime, RunDiscardTask, task.get());
+    if (ret == 0) {
+        task->SetId(timerId);
+        LOG(INFO) << "Schedule discard task success, taskid = " << task->Id();
+        std::lock_guard<bthread::Mutex> lk(mtx_);
+        unfinishedTasks_.emplace(timerId, std::move(task));
+        metric_->pending << 1;
+        return true;
+    }
+
+    LOG(ERROR) << "bthread_timer_add failed, ret = " << ret
+               << ", errno = " << errno << ", run task directly";
+    return false;
+}
+
+void DiscardTaskManager::Stop() {
+    std::unordered_set<bthread_timer_t> currentTasks;
+
+    {
+        std::lock_guard<bthread::Mutex> lk(mtx_);
+        for (auto& kv : unfinishedTasks_) {
+            currentTasks.emplace(kv.first);
+        }
+    }
```

Comment:
if didn't make a copy here, I have to hold the lock before the `for-loop`, and release it after the `for-loop`, and it will block other tasks that going to modify `unfinishedTasks_`.